% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/plot_region.R
\name{plot_region}
\alias{plot_region}
\alias{plot_region,NanoMethResult,character,numeric,numeric-method}
\alias{plot_region,NanoMethResult,factor,numeric,numeric-method}
\title{Plot region}
\usage{
plot_region(x, chr, start, end, ...)

\S4method{plot_region}{NanoMethResult,character,numeric,numeric}(
  x,
  chr,
  start,
  end,
  anno_regions = NULL,
  spaghetti = FALSE,
  span = NULL,
  window_prop = 0
)

\S4method{plot_region}{NanoMethResult,factor,numeric,numeric}(
  x,
  chr,
  start,
  end,
  anno_regions = NULL,
  spaghetti = FALSE,
  span = NULL,
  window_prop = 0
)
}
\arguments{
\item{x}{the NanoMethResult object.}

\item{chr}{the chromosome to plot.}

\item{start}{the start of the plotting region.}

\item{end}{the end of the plotting region.}

\item{...}{additional arguments.}

\item{anno_regions}{the data.frame of regions to be annotated.}

\item{spaghetti}{whether or not individual reads should be shown.}

\item{span}{the span for loess smoothing.}

\item{window_prop}{the size of flanking region to plot. Can be a vector of two
values for left and right window size. Values indicate proportion of gene
length.}
}
\value{
a patchwork plot containing the methylation profile in the specified
region.

a patchwork plot containing the methylation profile in the specified
region.
}
\description{
Plot region
}
\examples{
nmr <- load_example_nanomethresult()
plot_region(nmr, "chr7", 6703892, 6730431)

nmr <- load_example_nanomethresult()
plot_region(nmr, "chr7", 6703892, 6730431)

}
